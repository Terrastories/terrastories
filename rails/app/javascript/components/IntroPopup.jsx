import React, { PureComponent } from "react";

class IntroPopup extends PureComponent {

  constructor(props){
    super(props);
    this.state = { isPopped: true }
  }

  handleIntroPopup = () => {
    this.setState(prevState =>({
      isPopped: !prevState.isPopped
    }));
  }

  render(){
    return(
      <div className={this.state.isPopped ? 'intro-card isShown' : 'intro-card isHidden'}>
        <h2>{I18n.t("introduction")}</h2>
        <h3>{I18n.t("intro.question")}</h3>
        <p>{I18n.t("intro.explanation")}</p>

        <div className="intro-card--actions">
          <span className="count" onClick={this.handleIntroPopup}>{I18n.t("close")}</span>
        </div>
      </div>
    );
  }

}

export default IntroPopup;
