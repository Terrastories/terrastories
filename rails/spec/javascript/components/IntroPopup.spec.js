import React from 'react';
import { shallow } from 'enzyme';
import IntroPopup from 'components/IntroPopup';

describe("IntroPopup", () => {
    let props;
    let t_count;
    let component;

    beforeAll(() => {
        global.I18n = require('./__mocks__/I18n.mock.js');
        props = { isPopped: true };
        t_count = I18n.t.mock.calls.length;
        component = shallow(<IntroPopup {...props} />);
    });

    it("should render correctly", () => {
        expect(component).toMatchSnapshot();
        let divs = component.find("div");
        expect(divs.first().hasClass("intro-card")).toBe(true);
        expect(divs.last().hasClass("intro-card--actions")).toBe(true);
    });

    describe ("should use localized text:", () => {
        it("makes calls to I18n", () => {
            expect(I18n.t.mock.calls.length).toBeGreaterThan(t_count);
        });
        it("renders the localized text", () => {
            expect(I18n.t).toHaveBeenCalledWith("introduction");
            expect(I18n.t).toHaveBeenCalledWith("intro.question");
            expect(I18n.t).toHaveBeenCalledWith("intro.explanation");
            expect(I18n.t).toHaveBeenCalledWith("close");
        })
    });

    it("should initially pop up and then close when clicked", () => {
        let clickComponent = shallow(<IntroPopup {...props} />);
        let introCard = clickComponent.find("div.intro-card");
        expect(introCard.length).toBe(1);
        expect(introCard.hasClass("isShown")).toBe(true);
        expect(introCard.hasClass("isHidden")).toBe(false);
        clickComponent.find("div.intro-card--actions > span").simulate('click');
        /* IntroPopup should have re-rendered, so we need to re-find introCard */
        introCard = clickComponent.find("div.intro-card");
        expect(introCard.length).toBe(1);
        expect(introCard.hasClass("isShown")).toBe(false);
        expect(introCard.hasClass("isHidden")).toBe(true);
    });

    it("should default to popped up when no prop is given", () => {
        let defaultComponent = shallow(<IntroPopup />);
        expect(defaultComponent.state("isPopped")).toBe(true);
        let introCard = defaultComponent.find("div.intro-card");
        expect(introCard.length).toBe(1);
        expect(introCard.hasClass("isShown")).toBe(true);
    });

    //TODO: CODE REVIEW: Decide if this would be a future desired behavior, un-skip if so, otherwise delete
    describe.skip("should recognize the supplied props:", () => {
        describe("isPopped: {boolean}", () => {
            it("should display popped up when isPopped is true", () => {
                let popProps = {...props};
                popProps.isPopped = true;
                let poppedComponent = shallow(<IntroPopup {...popProps}/>);
                expect(poppedComponent.state("isPopped")).toBe(true);
                let introCard = poppedComponent.find("div.intro-card");
                expect(introCard.hasClass("isShown")).toBe(true);
            });
            it("should display hidden when isPopped is false", () => {
                let popProps = {...props};
                popProps.isPopped = false;
                let poppedComponent = shallow(<IntroPopup {...popProps}/>);
                expect(poppedComponent.state("isPopped")).toBe(false);
                let introCard = poppedComponent.find("div.intro-card");
                expect(introCard.hasClass("isHidden")).toBe(true);
            });
        });
    });
});