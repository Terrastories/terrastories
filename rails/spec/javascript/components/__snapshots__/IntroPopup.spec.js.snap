// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntroPopup should render correctly 1`] = `
<div
  className="intro-card isShown"
>
  <h2>
    introduction in local translation
  </h2>
  <h3>
    intro.question in local translation
  </h3>
  <p>
    intro.explanation in local translation
  </p>
  <div
    className="intro-card--actions"
  >
    <span
      className="count"
      onClick={[Function]}
    >
      close in local translation
    </span>
  </div>
</div>
`;
